src/permutation.jl

```diff
@@ -0,0 +1,65 @@
+export PermutationTest
+
+function ptstats(x,y,f)
+    xy = vcat(x,y)
+    rx = 1 : length(x)
+    ry = length(xy)-length(y)+1 : length(xy)
+    T = typeof(f(x))  # is there a better way?  e.g., does a funtype() exist, like eltype()??
+    (xy, rx, ry, T)
+end
+
+immutable PermutationTest <: HypothesisTests.HypothesisTest
+    observation
+    samples
+end
+
+"""
+    PermutationTest(x::Vector, y::Vector, f::Function [, n::Int])
+
+Perform a permutation test (a.k.a. randomization test) of the null hypothesis
+that the probability of `f(x)` is equal to `f(y)`.
+
+If `n` is *not* supplied, all possible permutations are sampled.  Otherwise,
+only `n` of the `factorial(length(vec(x,y)))` permutations are sampled.
+"""
+function PermutationTest{R<:Real,S<:Real}(x::AbstractVector{R}, y::AbstractVector{S}, f::Function)
+    xy, rx, ry, T = ptstats(x,y,f)
+    P = permutations(xy)
+    xsamples = Array(T,length(P))
+    ysamples = Array(T,length(P))
+    for (i,p) in enumerate(P)
+        xsamples[i] = f(p[rx])
+        ysamples[i] = f(p[ry])
+    end
+    PermutationTest(f(x) - f(y), xsamples - ysamples)
+end
+
+function PermutationTest{R<:Real,S<:Real}(x::AbstractVector{R}, y::AbstractVector{S}, f::Function, n::Int)
+    xy, rx, ry, T = ptstats(x,y,f)
+    xsamples = Array(T,n)
+    ysamples = Array(T,n)
+    for i = 1:n
+        shuffle!(xy)
+        xsamples[i] = f(xy[rx])
+        ysamples[i] = f(xy[ry])
+    end
```

Comment:
If you create views into `xy` outside the loop, then you don't need to do any indexing in the loop here, which will avoid allocation.